### ISSUE TITLE
Redis rate limiting does not work as expected

### ISSUE THREAD

**Issue by rphillipsz** [istio/istio#16167]
**Bug description**
I followed the example (https://raw.githubusercontent.com/istio/istio/release-1.2/samples/bookinfo/policy/mixer-rule-productpage-redis-quota-rolling-window.yaml) for setting up a rate limiter of 100 requests/second. Requests at well under this rate are being rejected.  Rates above this limit are being rejected, but not at the defined rate.

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[x] Policies and Telemetry
[ ] Security
[ ] Test and Release
[x] User Experience
[ ] Developer Infrastructure

**Expected behavior**
Request rates under the defined limit should never encounter a 429.  Rates above the limit should be restricted to roughly the defined limit.

**Steps to reproduce the bug**
Deploy the redis rate limiter with a limit of 100/sec with a 500ms rolling window.  Query the endpoint with a 0.4 second delay between requests.  Notice that you get frequent 429's.  Spam the endpoint at well over the rate and record the number of successful requests/second and notice that the pass rate is not close to 100/sec.

Rate limiter configuration:
```
apiVersion: config.istio.io/v1alpha2
kind: handler
metadata:
  name: redishandler
  namespace: istio-system
spec:
  compiledAdapter: redisquota
  params:
    redisServerUrl: <REDIS>:6379
    connectionPoolSize: 10
    quotas:
    - name: requestcountquota.instance.istio-system
      maxAmount: 100
      validDuration: 1s
      bucketDuration: 500ms
      rateLimitAlgorithm: ROLLING_WINDOW
---
apiVersion: config.istio.io/v1alpha2
kind: instance
metadata:
  name: requestcountquota
  namespace: istio-system
spec:
  compiledTemplate: quota
  params:
    dimensions:
      source: request.headers["x-forwarded-for"] | "unknown"
      destination: destination.labels["app"] | destination.workload.name | "unknown"
---
apiVersion: config.istio.io/v1alpha2
kind: QuotaSpec
metadata:
  name: request-count
  namespace: istio-system
spec:
  rules:
  - quotas:
    - charge: 1
      quota: requestcountquota
---
apiVersion: config.istio.io/v1alpha2
kind: QuotaSpecBinding
metadata:
  name: request-count
  namespace: istio-system
spec:
  quotaSpecs:
  - name: request-count
    namespace: istio-system
  services:
  - name: data-receiver-v3
    namespace: default
---
apiVersion: config.istio.io/v1alpha2
kind: rule
metadata:
  name: quota
  namespace: istio-system
spec:
  actions:
  - handler: redishandler
    instances:
    - requestcountquota
---
```
Observed query results with a 0.4 second delay between requests, a single thread, over 120 seconds:

```
25 errors of 200 requests (12% error rate) made in 120.644380 seconds.
1.66 requests/second.
Responses:
  200: 175 1.45 per second
  429: 24 0.20 per second
  502: 1 0.01 per second
```

Observed query results with no delay, 20 query threads, over 120 seconds:

```
16320 errors of 19750 requests (82% error rate) made in 120.232450 seconds.
164.27 requests/second.
Responses:
  200: 3430 28.53 per second
  429: 16262 135.25 per second
  502: 56 0.47 per second
  Post https://REDACTED/v1/data-receiver/metrics: net/http: request canceled (Client.Timeout exceeded while awaiting headers): 1 0.01 per second
```

**Version (include the output of `istioctl version --remote` and `kubectl version`)**
`istioctl version --remote`

```
client version: version.BuildInfo{Version:"1.1.3", GitRevision:"d19179769183541c5db473ae8d062ca899abb3be", User:"root", Host:"fbd493e1-5d72-11e9-b00d-0a580a2c0205", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.2-56-gd191797"}
citadel version: version.BuildInfo{Version:"b2915d396d159e54a88a259c971940348d8a9082", GitRevision:"b2915d396d159e54a88a259c971940348d8a9082", User:"root", Host:"75e030f278d1", GolangVersion:"go1.11.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.3-gke.0"}
galley version: version.BuildInfo{Version:"b2915d396d159e54a88a259c971940348d8a9082", GitRevision:"b2915d396d159e54a88a259c971940348d8a9082", User:"root", Host:"75e030f278d1", GolangVersion:"go1.11.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.3-gke.0"}
ingressgateway version: version.BuildInfo{Version:"b2915d396d159e54a88a259c971940348d8a9082", GitRevision:"b2915d396d159e54a88a259c971940348d8a9082", User:"root", Host:"75e030f278d1", GolangVersion:"go1.11.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.3-gke.0"}
pilot version: version.BuildInfo{Version:"b2915d396d159e54a88a259c971940348d8a9082", GitRevision:"b2915d396d159e54a88a259c971940348d8a9082", User:"root", Host:"75e030f278d1", GolangVersion:"go1.11.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.3-gke.0"}
policy version: version.BuildInfo{Version:"b2915d396d159e54a88a259c971940348d8a9082", GitRevision:"b2915d396d159e54a88a259c971940348d8a9082", User:"root", Host:"75e030f278d1", GolangVersion:"go1.11.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.3-gke.0"}
sidecar-injector version: version.BuildInfo{Version:"b2915d396d159e54a88a259c971940348d8a9082", GitRevision:"b2915d396d159e54a88a259c971940348d8a9082", User:"root", Host:"75e030f278d1", GolangVersion:"go1.11.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.3-gke.0"}
telemetry version: version.BuildInfo{Version:"b2915d396d159e54a88a259c971940348d8a9082", GitRevision:"b2915d396d159e54a88a259c971940348d8a9082", User:"root", Host:"75e030f278d1", GolangVersion:"go1.11.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.3-gke.0"}
```

`kubectl version`

```
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.1", GitCommit:"4485c6f18cee9a5d3c3b4e523bd27972b1b53892", GitTreeState:"clean", BuildDate:"2019-07-18T09:18:22Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.6-gke.13", GitCommit:"fcbc1d20b6bca1936c0317743055ac75aef608ce", GitTreeState:"clean", BuildDate:"2019-06-19T20:50:07Z", GoVersion:"go1.11.5b4", Compiler:"gc", Platform:"linux/amd64"}
```

**How was Istio installed?**
GKE deployment

**Environment where bug was observed (cloud vendor, OS, etc)**
GKE

Additionally, please consider attaching a [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) by attaching
the dump file to this issue.

**Comment by rphillipsz** [comment]
Are there any updates on this?

**Comment by gargnupur** [comment]
@rphillipsz : In the second case how many istio-policy pods do you have? 
Seems like https://github.com/istio/istio/issues/3028...

**Comment by rphillipsz** [comment]
In both cases, a single istio-policy pod.

**Comment by cluther** [comment]
I've isolated this issue to Istio 1.1.3 managed by GKE. This problem doesn't exist in Istio 1.3.0 deployed manually to GKE. The cause was issue #15543.

I believe this issue should be closed as a duplicate of #15543.

**Comment by gargnupur** [comment]
Thanks @cluther  for confirming. Closing the issue.
